### ISSUE TITLE
Handling of Runtime.json

### ISSUE THREAD

**Issue by forki** [fsprojects/Paket#1161]
originally from https://github.com/fsprojects/Paket/issues/714#issuecomment-150327458.

@ctaggart wrote:

"I think it would be good for Paket to look at the runtime.json files in the nuget packages to get a list of additional transitive dependencies for each platform.

In this example:
http://blog.ctaggart.com/2015/10/minimal-coreclr-console-app.html
Paket could look at the System.Console's runtime.json to see that it depends on runtime.win7.System.Console.

```
    {
      "runtimes": {
        "win7": {
          "System.Console": {
            "runtime.win7.System.Console": "4.0.0-beta-23420"
          }
        },
        "unix": {
          "System.Console": {
            "runtime.unix.System.Console": "4.0.0-beta-23420"
          }
        }
      }
    }
```

"

Why is this information not in the package meta data? Do we really need to download a package to see that there are other transitive dependencies?

@maartenba @yishaigalatzer

**Comment by forki** [comment]
another question is: when get this resolved?
at dev time I might be on win7, but want to deploy to linux? Other dev is on mac? What happens now?

**Comment by forki** [comment]
So I think this doesn't belong into the normal package resolution process, but needs to be added to a file in the coreclr project. It's basically meta data of the resulting coreclr project and the dnx runtime needs to resolve this on first run or something like that.

**Comment by forki** [comment]
But if this is true it would make real dependency management impossible.

**Comment by SamB** [comment]
@forki: I don't think the question of when to resolve these runtime dependencies has any kind of complete answer yet, but it would seem to depend on what you want to do with the package.

**DISCLAMER**: <small>I am more-or-less making most of this stuff up; I'm not sure I've ever even heard of `runtime.json` before.<small>

With respect to what you need just to hack on, build, and/or do basic testing of libraries and programs, depending on the role of a given dependency package:

  * **build tools**: These definitely need runtime dependencies resolved for the BUILD system.
  * **test runners**: You need these resolved for whichever HOST runtime(s) you want to test
  * **analyzers/fixers**: If your development environment supports these, you'll probably want their runtime deps resolved such that it can load them; whether or not it does, you probably want your compiler to use them.
  * **plain old libraries**: You want these resolved for whatever HOST runtime(s) you'll be messing about with

With respect to actually building distributable packages, it would seem to vary quite widely between package types:
  * For **NuGet** packages, the only added responsibilities you could reasonably have here are:
      * Get your *normal* deps right
      * If *your* code takes any runtime-specific deps, declare those
    Responsibility for actually resolving the runtime deps would then fall on whoever/whatever uses the package.
  * For **ZIP** ("xcopy") distributions, you'd have to resolve all runtime dependencies 
  * Debian packages would require all kinds of gymnastics (or at least patching/replacing most of the build scripts) that might seem like insanely overkill if you don't realize how hard it can be to make sure that the pre-compiled software you are shipping is truly free/open source software in both legal (I am allowed to modify this and use it for any purpose) and practical (I have the source code, I can rebuild the software to incorporate my changes) senses.
  * Other package types, like **MSI** or **VSIX**, would probably need *some* deps (runtime or otherwise) included in the package, and others delegated to other packages, and figuring out which is which would probably be up to a human.

TL;DR!

Given the lack of a simple answer to "when should runtime deps be resolved", and the fact that the answer depends both on the role of the package and on what one wants to do with it, Paket should allow developers to specify when to resolve runtime deps for different sets of packages and runtimes.  In particular, it should probably provide a function/command meaning essentially "please resolve runtime deps NOW for PACKAGES and RUNTIME".

**Comment by matthid** [comment]
This as been implemented in paket 5. It is disabled by default and can be enabled via the `PAKET_DISABLE_RUNTIME_RESOLUTION=false` environment variable.

We will need this information as soon as we replace nuget in netcore world to properly tell msbuild which files need to be published and similar stuff.
If some custom packages start to use this feature than FAKE and F# `#r "paket: nuget Test"`needs this information to download and load the required runtime assemblies for the current platform.
